lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp

```diff
@@ -1832,118 +1871,145 @@ void HdVP2Mesh::_UpdateDrawItem(
+                }
+            } else {
+                drawScene.setInstanceTransformArray(*renderItem, stateToCommit._instanceTransforms);
+            }
+
+            // upload any extra instance data
+            for (auto& entry : *primvarInfo) {
+                const MFloatArray& extraInstanceData = entry.second->_extraInstanceData;
+                if (extraInstanceData.length() == 0)
+                    continue;
+
+                const TfToken& primvarName = entry.first;
+                if (primvarName == HdVP2Tokens->displayColorAndOpacity) {
                     drawScene.setExtraInstanceData(
-                        *renderItem, kSolidColorStr, stateToCommit._instanceColors);
+                        *renderItem, kDiffuseColorStr, extraInstanceData);
                 }
             }
+
+            if (stateToCommit._instanceColors.length() == newInstanceCount * kNumColorChannels) {
+                drawScene.setExtraInstanceData(
+                    *renderItem, kSolidColorStr, stateToCommit._instanceColors);
+            }
+        }
 #if MAYA_API_VERSION >= 20210000
-            else if (newInstanceCount >= 1) {
+        else if (newInstanceCount >= 1) {
 #else
-            // In Maya 2020 and before, GPU instancing and consolidation are two separate systems
-            // that cannot be used by a render item at the same time. In case of single instance, we
-            // keep the original render item to allow consolidation with other prims. In case of
-            // multiple instances, we need to disable consolidation to allow GPU instancing to be
-            // used.
-            else if (newInstanceCount == 1) {
-                renderItem->setMatrix(&stateToCommit._instanceTransforms[0]);
-            } else if (newInstanceCount > 1) {
-                setWantConsolidation(*renderItem, false);
+        // In Maya 2020 and before, GPU instancing and consolidation are two separate systems
+        // that cannot be used by a render item at the same time. In case of single instance, we
+        // keep the original render item to allow consolidation with other prims. In case of
+        // multiple instances, we need to disable consolidation to allow GPU instancing to be
+        // used.
+        else if (newInstanceCount == 1) {
+            renderItem->setMatrix(&stateToCommit._instanceTransforms[0]);
+        } else if (newInstanceCount > 1) {
+            setWantConsolidation(*renderItem, false);
 #endif
-                drawScene.setInstanceTransformArray(*renderItem, stateToCommit._instanceTransforms);
+            drawScene.setInstanceTransformArray(*renderItem, stateToCommit._instanceTransforms);
+
+            if (stateToCommit._instanceColors.length() == newInstanceCount * kNumColorChannels) {
+                drawScene.setExtraInstanceData(
+                    *renderItem, kSolidColorStr, stateToCommit._instanceColors);
+            }
+
+            // upload any extra instance data
```

Comment:
Also need the upload in this case. The duplicated code here is a bit awkward but I didn't want to refactor this and risk breaking something I don't understand fully. This code path happens on the first frame when _usingInstancedDraw is not yet set.

Reply:
Ok. I think the duplication was due to a limitation of pre-2022 releases in which consolidation and GPU instancing cannot be active for a render item at the same time. For 2022, we can make the branch simpler. But yes, it is better not to refactor that part in the pull request.